src/main/java/org/takes/misc/VerboseList.java

```diff
@@ -0,0 +1,205 @@
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.takes.misc;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * Verbose List that wraps OutOfBoundsException with custom message.
+ *
+ * @author I. Sokolov (happy.neko@gmail.com)
+ * @version $Id$
+ * @param <T> Type of item
+ */
+@SuppressWarnings("PMD.TooManyMethods")
+public final class VerboseList<T> implements List<T> {
+
+    /**
+     * Original list.
+     */
+    private final transient List<T> origin;
+
+    /**
+     * Error message for IndexOutOfBoundsException.
+     */
+    private final transient String message;
+
+    /**
+     * Ctor.
+     * @param list Original list
+     * @param msg Error message for IndexOutOfBoundsException
+     */
+    public VerboseList(final List<T> list, final String msg) {
+        this.origin = list;
+        this.message = msg;
+    }
+
+    @Override
+    public int size() {
+        return this.origin.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return this.origin.isEmpty();
+    }
+
+    @Override
+    public boolean contains(final Object obj) {
+        return this.origin.contains(obj);
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return new VerboseIterator<T>(this.origin.iterator(), this.message);
+    }
+
+    @Override
+    public Object[] toArray() {
+        return this.origin.toArray();
+    }
+
+    @SuppressWarnings("PMD.UseVarargs")
+    @Override
+    public <E> E[] toArray(final E[] arr) {
+        return this.origin.toArray(arr);
+    }
+
+    @Override
+    public boolean add(final T item) {
+        return this.origin.add(item);
+    }
+
+    @Override
+    public boolean remove(final Object obj) {
+        return this.origin.remove(obj);
+    }
+
+    @Override
+    public boolean containsAll(final Collection<?> coll) {
+        return this.origin.containsAll(coll);
+    }
+
+    @Override
+    public boolean addAll(final Collection<? extends T> coll) {
+        return this.origin.addAll(coll);
+    }
+
+    @Override
+    public boolean addAll(final int index, final Collection<? extends T> coll) {
+        try {
+            return this.origin.addAll(index, coll);
+        } catch (final IndexOutOfBoundsException ex) {
+            throw new IllegalArgumentException(this.message, ex);
```

Comment:
@Happy-Neko Can't we still use `IndexOutOfBoundsException` with custom message? This will break usage of `List` interface for code that is not expecting `IllegalArgumentException`.